mtt/src/mtt.rs

```diff
@@ -0,0 +1,699 @@
+        }
+    }
+}
+
+/// Holds references to the platform allocated L2 and L2 tables in a Mutex.
+/// This should be constructed by calling init() with the platform allocated backing
+/// memory for the L2 and L1 tables.
+pub struct Mtt {
+    inner: Mutex<MttL2Directory>,
+}
+
+/// Exposes functions to change and query the type of memory ranges in the MTT.
+impl Mtt {
+    // Performs sanity checks on the platform L2 table
+    // TODO: Add additional sanity checks
+    fn validate_mtt(&self, l2_base_addr: u64) -> Result<()> {
+        let addr =
+            SupervisorPageAddr::new(RawAddr::supervisor(l2_base_addr)).ok_or(InvalidAddress)?;
+        if self.is_confidential(addr, L2_BYTE_SIZE)? {
+            Err(InvalidPlatformConfiguration)
+        } else {
+            Ok(())
+        }
+    }
+
+    /// Creates a Mtt instance.
+    /// The function performs some basic sanity checks on the passed-in parameters.
+    pub fn init(l2_base_addr: SequentialPages<InternalClean>) -> Result<Self> {
+        if l2_base_addr.length_bytes() as usize != L2_BYTE_SIZE {
+            return Err(InvalidPlatformConfiguration);
+        }
+        let l2_addr_bits = l2_base_addr.base().bits();
+        // L2 must be 8MB aligned
+        if l2_addr_bits & (L2_BYTE_SIZE as u64 - 1) != 0 {
+            return Err(InvalidPlatformConfiguration);
+        }
+        let mtt_l2_base: &'static mut [u64] = unsafe {
+            core::slice::from_raw_parts_mut(l2_addr_bits as *mut u64, MTT_L2_ENTRY_COUNT)
+        };
+        let instance = Self {
+            inner: Mutex::new(MttL2Directory { mtt_l2_base }),
+        };
+        instance.validate_mtt(l2_addr_bits)?;
+        Ok(instance)
+    }
+
+    fn is_range_of_type(
+        &self,
+        addr: SupervisorPageAddr,
+        len: usize,
+        memory_type: MttMemoryType,
+    ) -> Result<bool> {
+        let l2_dir = self.inner.lock();
+        let mut is_range_type_same = true;
+        {
+            let mut callback =
+                |addr: SupervisorPageAddr, _page_size: PageSize, l2_entry: MttL2Entry| {
+                    let region_memory_type = MttL2Directory::get_memory_type(addr, &l2_entry);
+                    if region_memory_type != memory_type {
+                        is_range_type_same = false;
+                        ControlFlow::Break(Ok(()))
+                    } else {
+                        ControlFlow::Continue(true)
+                    }
+                };
+
+            let mut mtt_range = MttRange::new(addr, len, l2_dir)?;
+            mtt_range.range_iterator(&mut callback)?;
+        }
+        Ok(is_range_type_same)
+    }
+
+    fn page_size_from_entry(l2_entry: &MttL2Entry) -> PageSize {
+        use MttL2Entry::*;
+        match l2_entry {
+            NonConfidential1G(_) | Confidential1G(_) => PageSize::Size1G,
+            Mixed2M(_) => PageSize::Size2M,
+            MttL2Entry::L1Entry(_) => PageSize::Size4k,
+        }
+    }
+
+    fn set_range_type(
```

Comment:
this, and everything downstream that actually sets the memory type, should be `unsafe`.